openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -103,20 +117,123 @@ class ParameterList(list):
+            will be initialized empty.
+        """
+        super().__init__()
+
+        # We keep track of the last parameter added as this will be
+        # used to set output units during serialization
+        self._last_added_param = None
+        input_parameter_list = input_parameter_list or []
+        for input_parameter in input_parameter_list:
+            self.append(input_parameter)
+            self._last_added_param = input_parameter_list
+
+    @property
+    def last_added_parameter(self):
+        """
+        Get a copy of the last parameter added to this ParameterList. Important
+        for serializing as the last parameter added determines the units used for
+        serializing all other parameters of this type.
+
+        Returns
+        -------
+        parameter : a ParameterType-derived object
+            The last parameter added to this ParameterList
+        """
+        return self._last_added_param
+
+    # def __setitem__(self, key, value):
+    #     """
+    #     Add a new entry to the ParameterList.
+    #     Parameters
+    #     ----------
+    #     key
+    #     value
+    #     """
+    #     super().__setitem__(key, value)
+    #     self._last_added_param = value
+
+    def append(self, parameter):
+        """
+        Add a ParameterType object to the end of the ParameterList
+
+        Parameters
+        ----------
+        parameter : a ParameterType-derived object
+
+        """
+        super().append(parameter)
+        self._last_added_param = parameter
+
+    def extend(self, other):
+        """
+        Add a ParameterType object to the end of the ParameterList
+
+        Parameters
+        ----------
+        parameter : a ParameterType-derived object
+
+        """
+        if not isinstance(other, ParameterList):
+            msg = 'ParameterList.extend(other) expected instance of ParameterList, ' \
+                  'but received {} (type {}) instead'.format(other, type(other))
+            raise TypeError(other)
+        super().extend(other)
+        if len(other) > 0:
+            self._last_added_param = other[-1]
+
+
+    def insert(self, index, parameter):
+        """
+        Add a ParameterType object as if this were a list
+
+        Parameters
+        ----------
+        parameter : a ParameterType-derived object
+
+        """
+        super().insert(index, parameter)
+        self._last_added_param = parameter
+
+    def __delitem__(self, item):
+        """
+        Delete item by index or SMIRKS
+        """
+        if type(item) is str:
+            # Try to find by SMIRKS
+            for parameter in self:
+                if parameter.smirks == item:
+                    self.remove(parameter)
+                    return
+
+        # Try numerical index access. This will grab the item to remove by index, and
+        # then call __delitem__ again on its own SMIRKS, finishing in the "if" statement above
+        item_to_remove = self[item].smirks
```

Comment:
I'm not positive, but I don't think I need to do this. If I run `del param_list[-1]`, the ParameterType will be removed from the list, BUT there will still be a pointer (or reference, or whatever python calls it) to it in the `_last_added_param` variable, so garbage collection should leave it alone... I'll go ahead and make a test for this to be sure.

Reply:
The `_last_added_param` does indeed stick around, even if the actual parameter is removed from the list. Test for this behavior added in 41000a9207cdc4430949459542138b650357f835

Reply:
I think the code will definitely still work - I just thought it seems a little odd that the collection will still keep a reference to (and use the value from) an item which was removed, which may (in some odd edge cases) lead to an unexpected user experience (but is most cases probably not!). Just a minor thing anyway!

Reply:
True. When I initially implemented this behavior, I had thought about having it get units from the penultimately-added parameter if the last-added parameter was deleted. But then, in a worst-case scenario, we'd need a data structure keeping track of the order in which all the parameters were added, in case a user kept deleting the last-added one. 

The whole units-from-the-last-added-parameter thing is not obviously correct, but I agree with John that it is the method that minimized "surprises" for the user.